torch/csrc/distributed/rpc/rpc_agent.cpp

```diff
@@ -6,16 +6,182 @@ namespace rpc {
 
 constexpr size_t WorkerInfo::MAX_NAME_LEN;
 
+// Large Time Duration for waiting on the condition variable until the map is
+// population. Cannot use
+// std::chrono::time_point<std::chrono::steady_clock>::max() due to a known
+// overflow-related bug.
+constexpr auto kLargeTimeDuration = std::chrono::hours(10000);
+
 RpcAgent::RpcAgent(
     WorkerInfo workerId,
     std::unique_ptr<RequestCallback> cb,
     std::chrono::milliseconds rpcTimeout)
     : workerInfo_(std::move(workerId)),
       cb_(std::move(cb)),
       rpcTimeout_(rpcTimeout),
-      profilingEnabled_(false) {}
+      profilingEnabled_(false),
+      rpcAgentRunning_(true) {
+  rpcRetryThread_ = std::thread(&RpcAgent::retryExpiredRpcs, this);
+}
+
+RpcAgent::~RpcAgent() {
+  shutdown();
```

Comment:
On further thought: Virtual method calls aren't such a great idea in base class destructors.
Can you remove this, and either call in the subclass, or (even better) expect the api user to call it?

The problem-
   1 the ~ProcessGroupAgent (or ThriftRpcAgent) destructor runs.
   2 the ProcessGroupAgent part of the class is effectively destroyed
   3  then ~RpcAgent destructor starts.
   4. It then calls back up into the ProcessGroupAgent::shutdown() method.
    5. This ends up maybe being ok for PGA because the super class implements idempotency. 
        But calling back up into a destroyed subclass's virtual functions tends to end badly.

Reply:
I'll add a guard to make `ThriftRpcAgent::shutdown()` idempotent. `test_duplicate_name` fails if we don't call `shutdown()` from the RpcAgent destructor. I did some logging to dig into this - the ProcessGroupAgent constructor throws an exception in `ProcessGroupAgent::collectNames()` since two agents have the same workerName. The python test asserts the exception is raised, then it calls the RpcAgent destructor and looks like it never calls the PGA destructor.

Reply:
- So, this is interesting - I agree that if you write a test program with the current version of clang, it appears to devirtualize the virtual shutdown() call() when called from ~RpcAgent's (maybe swapping the vtables) to avoid calling up to ProcessGroupAgent.
   Does anybody know offhand if this is in the spec, or simply undefined behavior? (though I'll argue below that either way we probably don't want it)

- That said, that means that the shutdown() call from ~RpcAgent() is always effectively broken - it will never call up to ProcessGroupAgent::shutdown(). The non ~RpcAgent shutdown() call is effectively always needed anyway, so there shouldn't be any harm from removing from the ~RpcAgent destructor.

- And honestly, in OSS code, we should be aiming for readability and avoiding error-prone patterns, rather than probing the boundaries of standard and potentially undefined behavior.
   If you Google search "calling virtual function from destructor", there isn't a magic link to the spec up top, but there are a ton of articles telling people to avoid doing this.

My advice here is to minimally avoid calling any virtuals from ~RpcAgent.
I'd also prefer we avoided from ~PGA an ~ThriftRpcAgent as well - though it seems to be much lower risk, since these classes are effectively implicitly final.

Reply:
According to https://wiki.sei.cmu.edu/confluence/display/cplusplus/OOP50-CPP.+Do+not+invoke+virtual+functions+from+constructors+or+destructors, we should not be calling virtual functions from constructors/destructors, since it won't result in calling the more derived implementation. I think @jjlilley is right that we do this in `ProcessGroupAgent` and ThriftRpcAgent currently, and ideally we should remove this pattern. 

@osalpekar, is there anyway you could modify `test_invalid_names` not construct/destruct RPC agents at all? I thought it didn't already but as per your logging looks like it does.

Reply:
`test_duplicate_names` essentially causes PGA to throw an exception in the constructor (due to a call to ProcessGroupAgent::collectNames()). Since this is in the constructor implmentation of the constructor (after the call to the RpcAgent constructor), the RpcAgent is constructed, but the PGA object is not. Thus on destruction, only the RpcAgent destructor is called, and the test fails (times out) if the destructor doesn't call shutdown since the retry thread is not joinable.

Reply:
I chatted with Pritam offline, and we decided a separate `RpcAgent::cleanup()` function would be the best way to proceed. It satisfies: 1) not calling a virtual function from the RpcAgent destructor 2) still works if PGA or Thrift Agents are not successfully constructed without depending on the child class's specific implementations 3) makes `RpcAgent::shutdown()` a pure virtual function and 4) doesn't require PGA or Thrift `shutdown()` implementations to call `RpcAgent::shutdown`.